packages/plugin-node-modules/sources/NodeModulesLinker.ts

```diff
@@ -0,0 +1,421 @@
+  locatorState += '\n__metadata:\n';
+  locatorState += '  version: 1\n';
+  for (const [locator, value] of locatorMap.entries()) {
+    locatorState += `\n"${locator}":\n`;
+    locatorState += `  size: ${value.size}\n`;
+    locatorState += `  locations:\n${Array.from(value.locations).map(loc => `    - "${loc}"\n`).join('')}`;
+  }
+  await xfs.writeFilePromise(locatorStatePath, locatorState);
+};
+
+const readLocatorState = async (locatorStatePath: PortablePath): Promise<NodeModulesLocatorMap> => {
+  const locatorMap: NodeModulesLocatorMap = new Map();
+  const locatorState = parseSyml(await xfs.readFilePromise(locatorStatePath, `utf8`));
+  delete locatorState.__metadata;
+  for (const [key, val] of Object.entries(locatorState)) {
+    locatorMap.set(key, {
+      size: val.size,
+      target: PortablePath.dot,
+      linkType: LinkType.HARD,
+      locations: new Set(val.locations),
+    });
+  }
+
+  return locatorMap;
+};
+
+interface LocationNode
+{
+  parentNode: LocationNode | null;
+  entries: Map<Filename, LocationNode>;
+}
+
+const persistNodeModules = async (rootPath: PortablePath, locatorMap: NodeModulesLocatorMap) => {
+  const NODE_MODULES = toFilename('node_modules');
+  const LOCATOR_STATE_FILE = toFilename('.yarn-state.yml');
+
+  const rootNmDirPath = ppath.join(rootPath, NODE_MODULES);
+  const locatorStatePath = ppath.join(rootNmDirPath, LOCATOR_STATE_FILE);
+
+  let prevLocatorMap: NodeModulesLocatorMap;
+  try {
+    prevLocatorMap = await readLocatorState(locatorStatePath);
+  } catch (e) {
+    prevLocatorMap = new Map();
+  }
+
+  const locationTree: LocationNode = {parentNode: null, entries: new Map()};
+  for (const val of prevLocatorMap.values()) {
+    for (const location of val.locations) {
+      const segments = location.split(ppath.sep).map(x => toFilename(x));
+      let node = locationTree;
+      for (const segment of segments) {
+        let childNode = node.entries.get(segment);
+        if (!childNode) {
+          childNode = {parentNode: node, entries: new Map()};
+          node.entries.set(segment, childNode);
+        }
+        node = childNode;
+      }
+    }
+  }
+
+  // Delete old locations
+  for (const [key, val] of prevLocatorMap.entries()) {
+    const newVal = locatorMap.get(key);
+    const locations = newVal ? newVal.locations : new Set();
+    for (const location of val.locations) {
+      if (!locations.has(location)) {
+        const segments = location.split(ppath.sep).map(x => toFilename(x));
+        let node = locationTree;
+        for (const segment of segments.slice(0, -1))
+          node = node.entries.get(segment)!;
+
+        let segCount = segments.length;
+        while (node.parentNode !== null && node.entries.size === 1) {
+          node = node.parentNode;
+          segCount--;
+        }
+        node.entries.delete(segments[segCount - 1]);
+        xfs.removeSync(ppath.join(rootPath, ...segments.slice(0, segCount)));
```

Comment:
I think you might need to handle the case where the target location has already been removed (for example because you removed a package A whose n_m contained one of the locations of package B).